examples/nextjs-with-typescript/src/Link.tsx

```diff
@@ -1,31 +1,50 @@
 /* eslint-disable jsx-a11y/anchor-has-content */
-import React from 'react';
+import * as React from 'react';
 import clsx from 'clsx';
 import { useRouter } from 'next/router';
 import NextLink, { LinkProps as NextLinkProps } from 'next/link';
 import MuiLink, { LinkProps as MuiLinkProps } from '@material-ui/core/Link';
 
 type NextComposedProps = React.AnchorHTMLAttributes<HTMLAnchorElement> & NextLinkProps;
 
-const NextComposed = React.forwardRef(function NextComposed(
-  props: NextComposedProps,
-  ref: React.Ref<any>,
-) {
-  const { as, href, prefetch, ...other } = props;
+const NextComposed = React.forwardRef<HTMLAnchorElement, NextComposedProps>((props, ref) => {
+  const { as, href, replace, scroll, passHref, shallow, prefetch, ...other } = props;
 
   return (
-    <NextLink href={href} prefetch={prefetch} as={as}>
+    <NextLink
+      href={href}
+      prefetch={prefetch}
+      as={as}
+      replace={replace}
+      scroll={scroll}
+      shallow={shallow}
+      passHref={passHref}
+    >
       <a ref={ref} {...other} />
     </NextLink>
   );
 });
 
 interface LinkPropsBase {
   activeClassName?: string;
+  innerRef?: React.Ref<HTMLAnchorElement>;
   naked?: boolean;
 }
 
-type LinkProps = LinkPropsBase & NextLinkProps & MuiLinkProps;
+type LinkProps = LinkPropsBase &
+  NextComposedProps &
+  Pick<
```

Comment:
Why do we need such a big Pick? How would big would the Omit be? It would automatically scale with new props whereas Pick doesn't.

Reply:
Omit would be about 260 props:

![image](https://user-images.githubusercontent.com/6209775/61305055-22223900-a7b8-11e9-8a6e-17af67817041.png)

Reply:
Ok so let's start a bit earlier: Why do we need to omit those props? It looks like we're spreading those to another component. If I look at the code I see no pick of those props but rather an omit of others so I would expect the types to reflect that behavior.

Reply:
If `Pick` is not used, there is a TypeScript error, which reads:

```
Type '((instance: unknown) => void) | RefObject<unknown> | (((instance: HTMLAnchorElement | null) => void) & ((instance: any) => void)) | (((instance: HTMLAnchorElement | null) => void) & RefObject<any>) | (RefObject<...> & ((instance: any) => void)) | (RefObject<...> & RefObject<...>) | null | undefined' is not assignable to type '((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | null | undefined'.
  Type 'RefObject<unknown>' is not assignable to type '((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | null | undefined'.
    Type 'RefObject<unknown>' is not assignable to type 'RefObject<HTMLAnchorElement>'.
      Type 'unknown' is not assignable to type 'HTMLAnchorElement'.ts(2322)
index.d.ts(93, 9): The expected type comes from property 'ref' which is declared here on type 'IntrinsicAttributes & AnchorHTMLAttributes<HTMLAnchorElement> & LinkProps & RefAttributes<HTMLAnchorElement>'
```
Not sure how to read this one. Would be awesome if you have any tips.

But, do consumers of this component really need all the 260+ props?
They can set any `AnchorHTMLAttributes` props, LinkProps from next.js. So all the basics (like className) are covered already. I was suggesting earlier reducing MuiLinkProps to a fewer properties (or exporting a new interface), like what next.js has:
https://github.com/zeit/next.js/blob/becd07402e6ed0145618972e80e81657050dadae/packages/next/client/link.tsx#L48-L57

If that was the case, picking could be avoided.

What are your thoughts?

Reply:
> But, do consumers of this component really need all the 260+ props?

Probably not but they're available on the actual link so I'd rather not deal with every single use case that might need some esoteric prop.

> Not sure how to read this one. Would be awesome if you have any tips.

Yeah that one is a current typescript limitation (see covariant vs. contravariant types). Since the `ref` is (currently) not passed through anyway by react we can omit that one. Hope that is sufficient?

Reply:
It is indeed sufficient